lib/src/client.dart

```diff
@@ -140,3 +141,48 @@ abstract class Client {
   /// do so can cause the Dart process to hang.
   void close();
 }
+
+const _clientToken = Object();
+
+/// Return the [Client] for the current [Zone], if one has been set.
+///
+/// NOTE: This function is explicitly hidden from the public API.
+Client? getZonedClient() => Zone.current[_clientToken] != null
+    ? Zone.current[_clientToken] as Client
+    : null;
+
+/// Runs [body] in its own [Zone] with [client] set as the default [Client].
+///
+/// For example:
+///
+/// ```
+/// class MyHttpClient extends http.BaseClient {
+///   @override
+///   Future<http.StreamedResponse> send(http.BaseRequest request) {
+///     // your implementation here
+///   }
+/// }
+///
+/// void main() {
+///  late Client client;
+///  if (Platform.isAndroid) {
+///    client = MyHttpClient();
+///  } else {
+///    client = Client();  // Use the default client.
+///  }
+///  runClientZoned(myFunction, client);
+/// }
+///
+/// myFunction() {
+///   // Uses the `Client` configured in `main`.
+///   final response = await get(Uri.parse("https://www.example.com/"));
+///   final client = Client();
+/// }
+/// ```
+///
+/// The [client] influences functions (e.g. `get`) and the [Client.new] factory.
+R runClientZoned<R>(R Function() body, Client client,
```

Comment:
I don't see a use case but that doesn't mean that there isn't one ;-)

You mean that performance of a single callback? It's hard to imagine that mattering. We would be introducing a tiny inconsistency with `HttpOverrides` but I don't think that matters either.

Reply:
It's potentially useful to pass in a factory instead of a single value. You can always pre-compute the value and pass in a function returning it every time, the other direction is not as interesting.

If we do that, we should/must do `Zone.current.bind(factory)` before storing the factory, so we are sure it's called in the parent zone.
That ensures that the user can have a factory like `() => someTest ? MyClient() : Client()` and not get an infinite recursion when that `Client()` is run in the child zone which goes right back and calls the factory again.

Reply:
Done and added a test.